pypeit/edgetrace.py

```diff
@@ -0,0 +1,4122 @@
+            binning (:obj:`str`, optional):
+                On-detector binning of the data ordered spectral then
+                spatial with format, e.g., `2,1`. Ignored if `img` is
+                an instance of :class:`pypeit.traceimage.TraceImage`.
+            save (:obj:`bool`, optional):
+                Save the result to the master frame.
+            debug (:obj:`bool`, optional):
+                Run in debug mode.
+            show_stages (:obj:`bool`, optional):
+                After ever stage of the auto trace, execute
+                :func:`show` to show the results. These calls to show
+                are always::
+
+                    self.show(thin=10, include_img=True, idlabel=True)
+
+        """
+        # Perform the initial edge detection and trace identification
+        self.initial_trace(img, bpm=bpm, det=det, binning=binning, save=False)
+        if show_stages:
+            self.show(thin=10, include_img=True, idlabel=True)
+
+        # Initial trace can result in no edges found
+        if not self.is_empty:
+            # Refine the locations of the trace using centroids of the
+            # features in the Sobel-filtered image.
+            self.centroid_refine()
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        # Initial trace can result in no edges found, or centroid
+        # refinement could have removed all traces (via `check_traces`)
+        if not self.is_empty:
+            # Fit the trace locations with a polynomial
+            self.fit_refine(debug=debug)
+            # Use the fits to determine if there are any discontinous
+            # trace centroid measurements that are actually components
+            # of the same slit edge
+            self.merge_traces(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        # Check if the PCA decomposition is possible; this should catch
+        # long slits
+        if self.can_pca():
+            # Use a PCA decomposition to parameterize the trace
+            # functional forms
+            self.pca_refine(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+            # Use the results of the PCA decomposition to rectify and
+            # detect peaks/troughs in the spectrally collapsed
+            # Sobel-filtered image, then use those peaks to further
+            # refine the edge traces
+            self.peak_refine(rebuild_pca=True, debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        elif self.par['sync_predict'] == 'pca':
+            # TODO: This causes the code to fault. Maybe there's a way
+            # to catch this earlier on?
+            msgs.error('Sync predict cannot use PCA because too few edges were found.  Either choose '+msgs.newline()+
```

Comment:
Added to the error message.  I originally it just change to `nearest` and go on, but this puts the pypeit file and the parameters actually used by the code out of sync.  This can cause the code to barf when you read in a saved EdgeTraceSet master file because I explicitly check that the parameters used to construct the file are the same as expected given the parameters passed in the instantiation of the object.  This is me being strict (and doesn't come up if you use the `from_file` method), but I think this is probably a case where we want to be strict.